src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+                    }
+
+                    builder.AppendLine();
+                }
+
+                return builder.ToString();
+            }
+#endif
+        }
+
+        /// <summary>
+        /// When we have a single sequence of `or` patterns, such as `A or B or C`
+        /// we can expand it to separate cases: `A`, `B` and `C`.
+        /// This composes. So a nested `or` sequence can also be expanded: `X and (A or B or C)`
+        /// can be expanded to `X and A`, `X and B` and `X and C`.
+        /// </summary>
+        private struct OrCases : IDisposable
+        {
+            public ArrayBuilder<(BoundPattern pattern, SyntaxNode? syntax)> Cases;
+
+            public OrCases()
+            {
+                Cases = ArrayBuilder<(BoundPattern pattern, SyntaxNode? syntax)>.GetInstance();
+            }
+
+            public void Add(BoundPattern pattern, SyntaxNode? syntax)
+            {
+                Cases.Add((pattern, pattern.WasCompilerGenerated ? null : syntax));
+            }
+
+            public void Dispose()
+            {
+                Cases.Free();
+            }
+        }
+
+        // The purpose of this rewriter is to push `not` patterns down the pattern tree.
+        // It needs to expand composite patterns in the process.
+        // TODO2 some patterns get erased (variable declarations?)
+        private class MoveNotPatternsDownRewriter : BoundTreeRewriterWithStackGuard
+        {
+            private static readonly MoveNotPatternsDownRewriter Instance = new MoveNotPatternsDownRewriter(negated: false);
+            private static readonly MoveNotPatternsDownRewriter NegatedInstance = new MoveNotPatternsDownRewriter(negated: true);
+
+            private readonly bool _negated;
+
+            public MoveNotPatternsDownRewriter(bool negated)
+            {
+                _negated = negated;
+            }
+
+            internal static BoundPattern Rewrite(BoundPattern pattern)
+            {
+                return (BoundPattern)Instance.Visit(pattern);
+            }
+
+            // A negated discard node represents an always false pattern
+            private bool IsNegatedDiscard(BoundPattern pattern, bool synthesized = false)
+            {
+                if (synthesized)
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern { WasCompilerGenerated: true } };
+                }
+                else
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern };
+                }
+            }
+
+            public override BoundNode? VisitBinaryPattern(BoundBinaryPattern node)
+            {
+                var resultLeft = (BoundPattern)Visit(node.Left);
+                var resultRight = (BoundPattern)Visit(node.Right);
+                var result = node.WithLeft(resultLeft).WithRight(resultRight);
+                return _negated ? result.WithDisjunction(!node.Disjunction) : result;
```

Comment:
>return _negated ? result.WithDisjunction(!node.Disjunction) : result; [](http://example.com/codeflow?start=16&length=69)

It feels like it will be more efficient to use ```Update``` once. It looks like with the current implementation we are allocating three new nodes even when no change to the original node is necessary.